[inductor] introduce kernel_inputs

coconutruben opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #156477
* #156292
* #156282
* #156281
* #156280
* __->__ #156279
* #156278
* #156277
* #156276

Summary:
# Why

- unified interface to wrap input_nodes and have a single implementation for common helpers
- introduce a surface to generate unique ids for specific inputs (shape, striding, etc)

# What

- kernel_inputs class, including an MM version
- replacing new call sites with that where appropriate

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

coconutruben committed:
[inductor] introduce kernel_inputs

Summary:
# Why

- unified interface to wrap input_nodes and have a single implementation for common helpers
- introduce a surface to generate unique ids for specific inputs (shape, striding, etc)

# What

- kernel_inputs class, including an MM version
- replacing new call sites with that where appropriate

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] introduce kernel_inputs"


Summary:
# Why

- unified interface to wrap input_nodes and have a single implementation for common helpers
- introduce a surface to generate unique ids for specific inputs (shape, striding, etc)

# What

- kernel_inputs class, including an MM version
- replacing new call sites with that where appropriate

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] introduce kernel_inputs"


Summary:
# Why

- unified interface to wrap input_nodes and have a single implementation for common helpers
- introduce a surface to generate unique ids for specific inputs (shape, striding, etc)

# What

- kernel_inputs class, including an MM version
- replacing new call sites with that where appropriate

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

NikhilAPatel reviewed (approved):
LGTM

masnesral commented on torch/_inductor/choices.py:
```diff
@@ -123,25 +124,34 @@ def get_mm_plus_mm_configs(
-        self, device_type: Optional[str] = "cuda"
+        self, kernel_inputs: MMKernelInputs
     ) -> partial[Generator[KernelTemplateParams, None, None]]:
+        device_type = (
```
nit: `device_type = kernel_inputs.device_type or "cuda"`
(Or maybe that's subjectively too opaque ;) )

masnesral commented on torch/_inductor/kernel_inputs.py:
```diff
@@ -0,0 +1,197 @@
+        Args:
+            input_nodes: A tuple of input nodes to store
+        """
+        self._input_nodes = input_nodes
```
nit: assert input_nodes not empty? (looks like code below expects there's always at least one element)

masnesral commented on torch/_inductor/kernel_inputs.py:
```diff
@@ -0,0 +1,197 @@
+        Get the device name information.
+
+        Returns:
+            A tuple of (gpu_name, vendor, model)
```
Comment looks out of sync with the return type?

masnesral commented on torch/_inductor/kernel_inputs.py:
```diff
@@ -0,0 +1,197 @@
+            A tuple of shape tuples with integer hints for each input node
+        """
+        import torch._inductor.config
+        from torch._inductor.virtualized import V
```
I guess you ran into a circular import?

coconutruben replied:
I actually don't think so, thanks for catching that. llm a bit too defensive in its coding

masnesral commented on torch/_inductor/kernel_inputs.py:
```diff
@@ -0,0 +1,197 @@
+        operands = self.nodes()[-2:]  # Get the last two nodes as operands
+        m = operands[0].get_size()[
+            0
+        ]  # M from first dimension of first operand (2nd to last)
```
nit: maybe move the comments a line above so the linter doesn't make it look this way lol

masnesral commented on torch/_inductor/kernel_inputs.py:
```diff
@@ -0,0 +1,197 @@
+        m = hinted_shapes[0][0]  # M from first dimension of first operand (2nd to last)
+        k = hinted_shapes[0][
+            1
+        ]  # K from second dimension of first operand (2nd to last)
```
nit: same comment about the linter making my eyes bleed

masnesral reviewed (approved):
some nits, but lgtm

coconutruben committed:
Update on "[inductor] introduce kernel_inputs"


Summary:
# Why

- unified interface to wrap input_nodes and have a single implementation for common helpers
- introduce a surface to generate unique ids for specific inputs (shape, striding, etc)

# What

- kernel_inputs class, including an MM version
- replacing new call sites with that where appropriate

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] introduce kernel_inputs"


Summary:
# Why

- unified interface to wrap input_nodes and have a single implementation for common helpers
- introduce a surface to generate unique ids for specific inputs (shape, striding, etc)

# What

- kernel_inputs class, including an MM version
- replacing new call sites with that where appropriate

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
